File: tobas_core/tobas_rqt_tools/tobas_rqt_tools/layouts/form_layout.py
from typing import Iterator, Tuple
from PyQt5.QtWidgets import QWidget, QFormLayout


class FormLayout(QFormLayout):
    """
    ===== QFormLayoutとの違い =====
    - イテレータを定義
    - 追加メソッド
    """

    def __iter__(self) -> Iterator[Tuple[QWidget, QWidget]]:
        for row in range(self.rowCount()):
            yield self.get_label(row), self.get_widget(row)

    def clear(self) -> None:
        """全てのフォームを削除する．"""
        while self.rowCount() > 0:
            self.removeRow(0)

    def get_label(self, row: int) -> QWidget:
        """指定した行のラベルを取得する．"""
        item = self.itemAt(row, QFormLayout.ItemRole.LabelRole)
        return item.widget()

    def get_widget(self, row: int) -> QWidget:
        """指定した行のウィジェットを取得する．"""
        item = self.itemAt(row, QFormLayout.ItemRole.FieldRole)
        return item.widget()

File: tobas_gui/tobas_qt_tools/src/layouts/form_layout.cpp
#include <QWidget>
#include <QLabel>
#include <QHBoxLayout>

#include "tobas_qt_tools/layouts/form_layout.hpp"

namespace qt
{
void FormLayout::addVAlignedRow(QWidget* label, QWidget* field)
{
  const auto label_layout = new QHBoxLayout();
  label_layout->addWidget(label, 0, Qt::AlignLeft | Qt::AlignVCenter);

  const auto label_widget = new QWidget();
  label_widget->setLayout(label_layout);

  addRow(label_widget, field);
}

void FormLayout::addVAlignedRow(const QString& label_text, QWidget* field)
{
  addVAlignedRow(new QLabel(label_text), field);
}

void FormLayout::clear()
{
  while (rowCount() > 0)
    removeRow(0);
}

QWidget* FormLayout::getLabel(int row)
{
  const auto item = itemAt(row, QFormLayout::LabelRole);
  return item->widget();
}

QWidget* FormLayout::getWidget(int row)
{
  const auto item = itemAt(row, QFormLayout::FieldRole);
  return item->widget();
}
}  // namespace qt
